cloud-storage/providers/local_folder/upload_file.c

```diff
@@ -0,0 +1,73 @@
+/* Copyright  (C) 2020 The RetroArch team
+ *
+ * ---------------------------------------------------------------------------------------
+ * The following license statement only applies to this file (upload_file.c).
+ * ---------------------------------------------------------------------------------------
+ *
+ * Permission is hereby granted, free of charge,
+ * to any person obtaining a copy of this software and associated documentation files (the "Software"),
+ * to deal in the Software without restriction, including without limitation the rights to
+ * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software,
+ * and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
+ * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
+ * WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+#include <file/file_path.h>
+#include <streams/file_stream.h>
+#include <string/stdstring.h>
+
+#include "../../cloud_storage.h"
+#include "local_folder_internal.h"
+
+#define BUFFER_SIZE 8192
+
+/* Copy a file from a RetroArch directory to a folder in the local folder provider
+ * base directory. Tries to create the destination file if it doesn't exist.
+ */
+bool cloud_storage_local_folder_upload_file(
+   cloud_storage_item_t *remote_dir,
+   cloud_storage_item_t *remote_file,
+   char *local_file)
+{
+   RFILE *src_file;
+   char *dest_filename;
+   RFILE *dest_file;
+   uint8_t buffer[BUFFER_SIZE];
+   uint8_t bytes_read;
+
+   src_file = filestream_open(local_file, RETRO_VFS_FILE_ACCESS_READ, RETRO_VFS_FILE_ACCESS_HINT_NONE);
+   if (!src_file)
+   {
+      return false;
+   }
+
+   /* Try to open the destination file, creating if needed */
+   dest_filename = pathname_join(remote_dir->id, remote_file->name);
+   dest_file = filestream_open(dest_filename, RETRO_VFS_FILE_ACCESS_WRITE, RETRO_VFS_FILE_ACCESS_HINT_NONE);
+   if (!dest_file) {
+      free(dest_filename);
+      filestream_close(src_file);
+      return false;
+   }
+   free(dest_filename);
+
+   /* Copy the file contents over 8k at a time */
+   bytes_read = filestream_read(src_file, buffer, BUFFER_SIZE);
```

Comment:
You can use do {} .. while (bytes_read > 0) perhaps if you think it's clearer.

Reply:
Hard to say what is better. I changed to a do..while loop. Only need one line to read bytes, but had to introduce an if statement.